File: Top-Interview-Questions-Leetcode/Easy-Collections/strings/longest-common-prefix.py
from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        n, length = len(strs), len(strs[0])
        
        for i in range(1, n):
            length = min(length, len(strs[i]))
            j, cur_length = 0, 0
            
            while j < length and strs[i][j] == strs[i - 1][j]:
                j += 1
                cur_length += 1
            
            length = cur_length
        
        return strs[0][:length]

def main():
    # Example usage
    solution = Solution()
    strs = ["flower", "flow", "flight"]
    result = solution.longestCommonPrefix(strs)
    print(result)  

if __name__ == "__main__":
    main()

File: Top-Interview-Questions-Leetcode/Easy-Collections/strings/longest-common-prefix.cpp
#include <iostream>
#include <string>
#include <vector>

class Solution {
public:
    std::string longestCommonPrefix(std::vector<std::string>& strs) {
        int n = strs.size();
        std::string longest_prefix = "";
        int maximum_size = INT_MAX;
        
        for (std::string str : strs) maximum_size = std::min(maximum_size, int(str.size()));
        
        for (int i = 0; i < maximum_size; ++i) {
            char ch = strs[0][i];
            bool check = 1;
            
            for (std::string str : strs) {
                if (str[i] != ch) {
                    check = 0;
                    break;
                }
            }
            
            if (check) longest_prefix += ch;
            else break;
        }
        
        return longest_prefix;
    }
    
    // Space optimized version
    std::string longestCommonPrefix_optimized(std::vector<std::string>& strs) {
        int n = strs.size();
        int len = strs[0].size();
        
        for (int i = 1; i < n; ++i) {
            len = std::min(len, (int)strs[i].size());
            
            int j = 0, cur = 0;
            
            while (j < len and strs[i][j] == strs[i - 1][j]) j++, cur++;
            
            len = cur;
        }
        
        return strs[0].substr(0, len);
    }
};

int main() {
    Solution solution;
    std::vector<std::string> strs = {"flower", "flow", "flight"};
    std::string result = solution.longestCommonPrefix(strs);
    std::cout << "Longest common prefix: " << result << std::endl; 
    return 0;
}   